go.mod

```diff
@@ -76,38 +81,51 @@ require (
-	github.com/spf13/cobra v1.4.0 // indirect
+	github.com/spf13/cobra v1.5.0 // indirect
 	github.com/spf13/pflag v1.0.5 // indirect
 	github.com/xlab/treeprint v0.0.0-20181112141820-a009c3971eca // indirect
 	go.starlark.net v0.0.0-20200306205701-8dd3e2ee1dd5 // indirect
 	golang.org/x/mod v0.6.0-dev.0.20220106191415-9b9b3d81d5e3 // indirect
-	golang.org/x/oauth2 v0.0.0-20220411215720-9780585627b5 // indirect
-	golang.org/x/sys v0.0.0-20220209214540-3681064d5158 // indirect
+	golang.org/x/oauth2 v0.0.0-20220608161450-d0670ef3b1eb // indirect
+	golang.org/x/sys v0.0.0-20220520151302-bc2c85ada10a // indirect
 	golang.org/x/term v0.0.0-20210927222741-03fcf44c2211 // indirect
 	golang.org/x/text v0.3.7 // indirect
 	golang.org/x/time v0.0.0-20220210224613-90d013bbcef8 // indirect
 	golang.org/x/tools v0.1.10-0.20220218145154-897bd77cd717 // indirect
-	golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1 // indirect
 	gomodules.xyz/jsonpatch/v2 v2.2.0 // indirect
 	google.golang.org/appengine v1.6.7 // indirect
-	google.golang.org/protobuf v1.27.1 // indirect
+	google.golang.org/protobuf v1.28.0 // indirect
 	gopkg.in/inf.v0 v0.9.1 // indirect
 	gopkg.in/yaml.v2 v2.4.0 // indirect
-	gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b // indirect
-	k8s.io/apiextensions-apiserver v0.24.1 // indirect
+	gopkg.in/yaml.v3 v3.0.1 // indirect
+	k8s.io/apiextensions-apiserver v0.24.2
 	k8s.io/cli-runtime v0.24.1 // indirect
-	k8s.io/component-base v0.24.1 // indirect
+	k8s.io/component-base v0.24.2 // indirect
 	k8s.io/kube-openapi v0.0.0-20220328201542-3ee0da9b0b42 // indirect
 	k8s.io/kubectl v0.24.1 // indirect
 	k8s.io/utils v0.0.0-20220210201930-3a6ce19ff2f9 // indirect
+	sigs.k8s.io/cluster-api v1.2.2
 	sigs.k8s.io/json v0.0.0-20211208200746-9f7c6b3444d2 // indirect
 	sigs.k8s.io/kustomize/api v0.11.4 // indirect
 	sigs.k8s.io/kustomize/kyaml v0.13.6 // indirect
 	sigs.k8s.io/structured-merge-diff/v4 v4.2.1 // indirect
 )
 
+require (
+	github.com/go-logr/zapr v1.2.0 // indirect
+	github.com/nxadm/tail v1.4.8 // indirect
+	go.uber.org/atomic v1.7.0 // indirect
+	go.uber.org/multierr v1.6.0 // indirect
+	go.uber.org/zap v1.19.1 // indirect
+	golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1 // indirect
+	gopkg.in/check.v1 v1.0.0-20201130134442-10cb98267c6c // indirect
+	gopkg.in/tomb.v1 v1.0.0-20141024135613-dd632973f1e7 // indirect
+)
+
 replace (
 	github.com/metal3-io/baremetal-operator => github.com/openshift/baremetal-operator v0.0.0-20220510135004-05f60709e46d
 	github.com/metal3-io/baremetal-operator/apis => github.com/openshift/baremetal-operator/apis v0.0.0-20220510135004-05f60709e46d
 	github.com/metal3-io/baremetal-operator/pkg/hardwareutils => github.com/openshift/baremetal-operator/pkg/hardwareutils v0.0.0-20220510135004-05f60709e46d
+	github.com/metal3-io/cluster-api-provider-metal3 => github.com/metal3-io/cluster-api-provider-metal3 v1.2.1-0.20220928094713-6d07ee63f3aa // release-1.2 branch
```

Comment:
I do not understand, can you elaborate please? 🤔

Reply:
`openshift/cluster-api-provider-baremetal`, and `metal3-io/cluster-api-provider-metal3` are the same project separated by years of changes: I don't think we want to depend on a future version of ourselves.

Reply:
I have no good idea how to resolve this:

- I need the metal3 remediation API here and in the MAPI project: here for the controllers, in MAPI for setting up the webhooks
- Copying the API here would result in a circular dependency: CAPBM dependens on MAPI, MAPI would then need to depend on CAPBM
- To my best knowledge the webhooks can't be configured here, because CAPBM isn't a cluster operator. So I can't remove the API dependency at MAPI.
- Copying the API to MAPI also doesn't make makes IMHO, that would be even more code there which doesn't belong there (next to the webhook configs...) but here.

Do you have an idea?

Reply:
@honza Have you got any suggestions on how to move forward here?  What would your team consider the best approach?

Reply:
I think it's fine if `openshift/machine-api-operator`, and `openshift/cluster-api-provider-baremetal` both depend on `metal3-io/cluster-api-provider-metal3/api`. If you need anything that is contained in `metal3-io/cluster-api-provider-metal3` here, you will need to backport it.  It looks like you're only using the `api` "submodule", so I'm not sure why the `go.mod` here has that line.

I think this should solve the problem for you if I understand correctly.

Reply:
> I think it's fine if openshift/machine-api-operator, and openshift/cluster-api-provider-baremetal both depend on metal3-io/cluster-api-provider-metal3/api

great

> so I'm not sure why the go.mod here has that line

will check, can very well be that's it's a leftover

Reply:
PR is rebased, and I run go mod tidy and vendor. It was a leftover indeed.

@honza can you please confirm that you are ok with the general approach of this PR (and the MAO counterpart)? I will ask the cloud team to merge the MAO PR then, which will resolve dependency issues of the e2e tests on this PR. I will have a deeper look at the e2e results afterwards and address the other comments on this PR. Thanks.

Reply:
@slintes Yes

Reply:
cool, thanks!